examples/export-lora/export-lora.cpp

```diff
@@ -1,465 +1,403 @@
-            if (++i >= argc) {
-                invalid_param = true;
-                break;
-            }
-            params->n_threads = std::stoi(argv[i]);
-            if (params->n_threads <= 0) {
-                params->n_threads = std::thread::hardware_concurrency();
-            }
-        } else if (arg == "-h" || arg == "--help") {
-            export_lora_print_usage(argc, argv, &default_params);
-            exit(0);
-        } else {
-            fprintf(stderr, "error: unknown argument: '%s'\n", arg.c_str());
-            export_lora_print_usage(argc, argv, &default_params);
-            exit(1);
+        auto general_arch_base = get_kv_str(base_model.ctx_gguf, "general.architecture");
+        auto general_arch_lora = get_kv_str(adapter->ctx_gguf,   "general.architecture");
+        if (general_arch_base != general_arch_lora) {
+            throw std::runtime_error("model arch and LoRA arch mismatch");
         }
     }
 
-    if (params->fn_model_base == default_params.fn_model_base) {
-        fprintf(stderr, "error: please specify a filename for model-base.\n");
-        export_lora_print_usage(argc, argv, &default_params);
-        exit(1);
-    }
-    if (params->fn_model_out == default_params.fn_model_out) {
-        fprintf(stderr, "error: please specify a filename for model-out.\n");
-        export_lora_print_usage(argc, argv, &default_params);
-        exit(1);
-    }
-    if (invalid_param) {
-        fprintf(stderr, "error: invalid parameter for argument: '%s'\n", arg.c_str());
-        export_lora_print_usage(argc, argv, &default_params);
-        exit(1);
+    ggml_type get_out_tensor_type(struct ggml_tensor * t) {
+        if (t->type == GGML_TYPE_F32) {
+            return GGML_TYPE_F32;
+        } else {
+            return GGML_TYPE_F16;
+        }
     }
-    return true;
-}
 
-static void free_lora(struct lora_data * lora) {
-    if (lora->ctx != NULL) {
-        ggml_free(lora->ctx);
-    }
-    delete lora;
-}
+    void run_merge() {
+        // prepare metadata
+        gguf_set_kv(ctx_out, base_model.ctx_gguf);
+        // output is forced to f16 for now
+        gguf_set_val_u32(ctx_out, "general.file_type", LLAMA_FTYPE_MOSTLY_F16);
+
+        // check if all lora adapters have the same tensors
+        // TODO: remove this when we can support merging subset of adapters. Ref: https://github.com/ggerganov/llama.cpp/pull/8607#discussion_r1686027777
+        if (adapters.size() > 1) {
+            auto & base_adapter = adapters[0];
+            for (size_t i = 1; i < adapters.size(); ++i) {
+                for (auto & it : base_adapter->tensors) {
+                    if (base_adapter->get_tensor(it.first) == nullptr) {
+                        throw std::runtime_error("Input adapters do not have the same list of tensors. This is not yet supported. Please merge the adapter one-by-one instead of merging all at once.");
+                    }
+                }
+            }
+        }
```

Comment:
I don't think that this does anything as it is. Changing it to iterate over the tensors of `adapters[i]`, and checking that the number of tensors in all the adapters is the same to prevent accepting loras that have less tensors than the first one, may work.

Reply:
The current logic make sure that all tensors in `adapter[i]` exists in `adapter[0]`. But yeah there can be also the case where `adapter[i]` has more tensors than `adapter[0]`. So I think adding a tensor count check here is the easy fix.

Reply:
I attempted my fix on https://github.com/ggerganov/llama.cpp/pull/8607/commits/4bd55ec02c4039a2ec7eec36ce8869c239289058

Reply:
I might be missing something, but I am assuming that `adapter->get_tensor` finds a tensor in `adapter->tensors`, so it will always be true here since it is iterating over the tensors of the same adapter that it checks. I am assuming that the intention is to check the tensors of each adapter against the `base_adapter`.